In [64]:
from random import normalvariate, random
from itertools import count, groupby, islice
from datetime import date
import math

# 1.  `read_fake_data`
Let's explore how the `read_fake_data` function works.
* It returns a generator.
* This generator can be called via `__next__()`
* Another way to call it successively is via `islice`. This is essentially a `for` loop, calling `__next__()` on the generator.

In [50]:
def read_fake_data(filename, debug=False):
    """
    This function returns a generator.
    Example usage:
       gen = read_fake_date("test_file_name")
       a = gen.__next__()
       => a = (1, 0.98234)

    Note: docs on count -> https://docs.python.org/3/library/itertools.html#itertools.count
    Essentially a wrapper around a basic while True loop
    """
    for i in count():
        if debug:
            print(f"Yielding data point #{i} from read_fake_data generator.")
        sigma = random() * 10
        day = date.fromtimestamp(i)
        value = normalvariate(0, sigma)
        yield (day, value)

In [51]:
data = read_fake_data("test")

In [52]:
data_point_1 = data.__next__()
print(data_point_1)

(datetime.date(1969, 12, 31), -6.514989488335621)


In [53]:
data_point_2 = data.__next__()
print(data_point_2)

(datetime.date(1969, 12, 31), -0.00857880210504717)


In [54]:
data_point_3_through_13 = islice(data, 5)
for point in list(data_point_3_through_13):
    print(point)

(datetime.date(1969, 12, 31), 3.293737797456721)
(datetime.date(1969, 12, 31), -2.201777677807986)
(datetime.date(1969, 12, 31), 1.2134248426474705)
(datetime.date(1969, 12, 31), 0.4233273581294428)
(datetime.date(1969, 12, 31), -8.169808005133275)


In [55]:
for i in range(0, 5):
    print(data.__next__())

(datetime.date(1969, 12, 31), 2.449537733222146)
(datetime.date(1969, 12, 31), 3.343676823220588)
(datetime.date(1969, 12, 31), 2.1597825522991054)
(datetime.date(1969, 12, 31), -3.6896542106219634)
(datetime.date(1969, 12, 31), -1.9765936516188247)


# 2. `day_grouper`

In [56]:
def day_grouper(iterable):
    """
    lambda takes in a data point of form: time, value
    Returns itertools groupby, which is an iterator with a next method

    Note: date.fromtimestamp() will return (1969, 12, 31) for any value in range [0, 25199]
    Hence, the first 25200 pieces of read fake data will be grouped together for day (1969, 12, 31)
    """
    key = lambda timestamp_value: timestamp_value[0]
    return groupby(iterable, key)

Let's just look at the standard `itertools.groupby` to start (note, contents must be presorted):

In [203]:
iterable = ["a", "a", "a", "a", "b", "b", "b", "b", "c", "c"]
grouped = groupby(iterable)

for group_name, group in grouped:
    print("Group Name:", group_name, "  Group contents:", list(group))

Group Name: a   Group contents: ['a', 'a', 'a', 'a']
Group Name: b   Group contents: ['b', 'b', 'b', 'b']
Group Name: c   Group contents: ['c', 'c']


Above, `groupby` takes in an iterable that is simple a precomputed list. Could it take in a generator? Yes! Let's look at a generator that creates the above list:

In [337]:
def simple_generator(debug=True):
    i = 0
    while True:
        if debug:
            print("Generator hit and yielding a value")
        if i < 4:
            yield "a"
        elif i < 8:
            yield "b"
        elif i < 11:
            yield "c"
        else: 
            break
        i += 1

In [338]:
iterable = simple_generator()

In [339]:
first_iterable_value = iterable.__next__()
print(first_iterable_value)

Generator hit and yielding a value
a


In [340]:
iterable = simple_generator()
print(
    list(islice(iterable, 10))
)

Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
['a', 'a', 'a', 'a', 'b', 'b', 'b', 'b', 'c', 'c']


Can `groupby` handle a generator as an input (as compared to a simple precomputed list?). The answer is yes! We can see this in a variety of ways. To start, let's see that `groupby` returns two items: the group `key` and an `iterator`. This means that `groupy` by does _not_ precompute the groups and group contents! It must be accessed via `__next__()`. This is great since it gives us more control and prevents loading all contents of generator into memory unless we want to. Let's take a look at how it can be accessed:

In [359]:
iterable = simple_generator(debug=True)
grouped = groupby(iterable)

In [360]:
for group_name, group in grouped:
    print("Group Name:", group_name)
    while True:
        try:
            print(f"Group {group_name} value: ", group.__next__())
        except StopIteration:
            print("\n")
            break

Generator hit and yielding a value
Group Name: a
Group a value:  a
Generator hit and yielding a value
Group a value:  a
Generator hit and yielding a value
Group a value:  a
Generator hit and yielding a value
Group a value:  a
Generator hit and yielding a value


Group Name: b
Group b value:  b
Generator hit and yielding a value
Group b value:  b
Generator hit and yielding a value
Group b value:  b
Generator hit and yielding a value
Group b value:  b
Generator hit and yielding a value


Group Name: c
Group c value:  c
Generator hit and yielding a value
Group c value:  c
Generator hit and yielding a value
Group c value:  c
Generator hit and yielding a value




We can also grab _all_ of the contents by advancing the iterator entirely to the end via `list`. This looks like:

In [361]:
iterable = simple_generator(debug=True)
grouped = groupby(iterable)

for group_name, group in grouped:
    print("Group Name:", group_name, "  Group contents:", list(group))

Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Group Name: a   Group contents: ['a', 'a', 'a', 'a']
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Group Name: b   Group contents: ['b', 'b', 'b', 'b']
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Group Name: c   Group contents: ['c', 'c', 'c']


How does this work in the context of our example? Let's take a look via the `day_grouper` function and our `read_fake_data` function:

In [369]:
data = read_fake_data("test")
data_day = day_grouper(data)

In [370]:
iter(data_day)

Remember, `data` is a generator and `data_day` is an `itertools.groupby` that is taking `data` as it's input iterable. In other words, `data_day` is a groupby that has taken a generator as an input. groupby returns an `itertools.groupby`, which is also an iterable:

In [371]:
type(iter(data_day)) == type(data_day)

True

So, the `groupby` will be successively calling `__next__()` on the generator (an iterable), which will be yielding day/value tuple pairs:

In [372]:
group_name, group_iterable = data_day.__next__()

In [373]:
print(group_name)
print(group_iterable)

1969-12-31


In [374]:
print(group_iterable.__next__())

(datetime.date(1969, 12, 31), -1.2140982362813253)


In [375]:
print(group_iterable.__next__())

(datetime.date(1969, 12, 31), -8.435549216466036)


In [376]:
print(group_iterable.__next__())

(datetime.date(1969, 12, 31), 2.5887059614629875)


In [377]:
print(group_iterable.__next__())

(datetime.date(1969, 12, 31), 8.36185095178034)


In [379]:
for pair in list(islice(group_iterable, 10)): 
    print(pair)

(datetime.date(1969, 12, 31), 12.858924260744535)
(datetime.date(1969, 12, 31), -8.102339986813885)
(datetime.date(1969, 12, 31), 1.1281637632800638)
(datetime.date(1969, 12, 31), 2.616516481219527)
(datetime.date(1969, 12, 31), 0.5825416034689577)
(datetime.date(1969, 12, 31), -5.475462044963013)
(datetime.date(1969, 12, 31), 1.734426388223523)
(datetime.date(1969, 12, 31), 2.8339230931639454)
(datetime.date(1969, 12, 31), -0.4247739211417971)
(datetime.date(1969, 12, 31), 9.731967617917219)


Above, `group_iterable` contains all of the day/value tuple's for day (`group_name`) `datetime.date(1969, 12, 31)`. We can see that when `group_iterable.__next__()` is called, our `read_fake_data` yields a value by enabling `debug`:

In [397]:
data = read_fake_data("test", debug=True)
data_day = day_grouper(data)

In [398]:
group_name, group_iterable = data_day.__next__()

Yielding data point #0 from read_fake_data generator.


In [400]:
group_iterable.__next__()

Yielding data point #1 from read_fake_data generator.


(datetime.date(1969, 12, 31), -3.0773053343360584)

In [401]:
for pair in list(islice(group_iterable, 10)): 
    print(pair)

Yielding data point #2 from read_fake_data generator.
Yielding data point #3 from read_fake_data generator.
Yielding data point #4 from read_fake_data generator.
Yielding data point #5 from read_fake_data generator.
Yielding data point #6 from read_fake_data generator.
Yielding data point #7 from read_fake_data generator.
Yielding data point #8 from read_fake_data generator.
Yielding data point #9 from read_fake_data generator.
Yielding data point #10 from read_fake_data generator.
Yielding data point #11 from read_fake_data generator.
(datetime.date(1969, 12, 31), -2.438625884470149)
(datetime.date(1969, 12, 31), 1.4012717712493605)
(datetime.date(1969, 12, 31), 4.99099942687332)
(datetime.date(1969, 12, 31), -10.853344799696204)
(datetime.date(1969, 12, 31), -4.363478914024846)
(datetime.date(1969, 12, 31), 1.157289121329716)
(datetime.date(1969, 12, 31), 0.8902768894172846)
(datetime.date(1969, 12, 31), -2.7388773546196497)
(datetime.date(1969, 12, 31), -1.4049422890990317)
(datetim

One of the most important things to keep in mind with the `groupby` is that there are technically 2 iterables that we are dealing with. 
1. `data_day` is an iterable. Every time we call `data_day.__next__()` we are going to have the generator yield values, until the `key` no longer matches for the group. Again, this looks like:

In [445]:
iterable = simple_generator(debug=True)
grouped = groupby(iterable)

In [446]:
a_group = grouped.__next__()
print(a_group)

Generator hit and yielding a value
('a', <itertools._grouper object at 0x102b5f1d0>)


We are now in the `a` group. `groupby` essentially allows `__next__()` to be called on `grouped`, and it will keep requesting values from the generator until the group no longer matches. At that point, it will return the new group, in this case `b`:

In [447]:
grouped.__next__()

Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value


('b', <itertools._grouper at 0x102d5f128>)

As expected, `b` is returned. If we call it again, `groupby` will continue requesting values from the generator until the `b` group is no longer matched (we expect to see 4 values), in this case returning the `c` group:

In [448]:
grouped.__next__()

Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value


('c', <itertools._grouper at 0x102b5f7f0>)

And calling `__next__()` a final time will yield the values from the generator where the group is `c`. This should be 3, and we should expect to stop at this point and have `None` returned, since the generator has been exhausted (meaning a `StopIteration` will be thrown):

In [449]:
grouped.__next__()

Generator hit and yielding a value
Generator hit and yielding a value
Generator hit and yielding a value


StopIteration: 

Note, to get a better understanding of generator _state_ let's look at the `a_group` tuple that was returned earlier: `('a', <itertools._grouper object at 0x102b5f1d0>)`. Can we call `__next__()` on the iterator that was returned, `a[0]`? 

In [452]:
a_group[1].__next__()

StopIteration: 

No, we cannot! The generator (`simple_generator`) has been exhausted and has no more values to yield. The way that `groupby` works is that it will grab a value from the generator/iterable and start a new group. This looks like:

```
grouped = groupby(iterable)
a_group = grouped.__next__()
```

This group that is returned, `a_group` above, will have a group name and iterable returned (that will gather group values). Via `__next__()`, `a_group[1]` will then continue to grab values from the generator/iterable where the group is present, and finally throw a `StopIteration` error when the group does not match. This causes a break to occur. Note, that **this is exhausting values from the generator**. In other words, the generator state is being updated and we cannot gather these values again at a later point in time. This can be seen in detail below:

In [453]:
iterable = simple_generator(debug=True)
grouped = groupby(iterable)

In [454]:
a_group = grouped.__next__()

Generator hit and yielding a value


In [456]:
a_group[1].__next__()

'a'

In [457]:
a_group[1].__next__()

Generator hit and yielding a value


'a'

In [458]:
a_group[1].__next__()

Generator hit and yielding a value


'a'

In [459]:
a_group[1].__next__()

Generator hit and yielding a value


'a'

In [460]:
a_group[1].__next__()

Generator hit and yielding a value


StopIteration: 

We see that four times `Generator hit and yielding a value` is printed, meaning out generator is hit and yields a value. The final time, it is hit, but the value is `b` and our group is `a`, so a `StopIteration` error is thrown. 

# 3. `check_anomaly`
How does `check_anomaly` fit into all of this? Well, let's start by passing it a single `day_data_tuple` as input:

In [18]:
def check_anomaly(day_data_tuple):
    """
    Find mean, std, and maximum values for the day. Using a single pass (online)
    mean/std algorithm allows us to only read through the day's data once.

    Note: M2 = 2nd moment, variance
          day_data is an iterable, returned from groupby, and we request values via for loop
    """
    (day, day_data) = day_data_tuple

    n = 0
    mean = 0
    M2 = 0
    max_value = 0
    for timestamp, value in day_data:
        n += 1
        delta = value - mean
        mean = mean + (delta / n)
        M2 += delta * (value - mean)
        max_value = max(max_value, value)
    variance = M2 / (n - 1)
    standard_deviation = math.sqrt(variance)

    # Check if day's data is anomalous, if True return day
    if max_value > mean + 6 * standard_deviation:
        return day
    return False

What does `day_data_tuple` look like? Well, it is essentially a single group with the key being the day and all of it's associated values. Specifically it a tuple, shown below as `day_1_tuple`. It is returned from `day_grouper`. It looks like:

In [485]:
data = read_fake_data("test", debug=True)
data_day = day_grouper(data)

In [486]:
day_1_tuple = data_day.__next__()

Yielding data point #0 from read_fake_data generator.


In [489]:
day_1_tuple

(datetime.date(1969, 12, 31), <itertools._grouper at 0x102b5fb70>)

At index 1 we see that `day_1_tuple` has an iterator. This iterator will request values from our fake data generator, continually yielding them until the day no longer matches the group date (`datetime.date(1969, 12, 31)` above). These values can be accessed via a `for` loop, shown below. Note, `i` is only used to avoid printing 25000+ values:

In [488]:
i = 0
for timestamp, value in day_1_tuple[1]:
    if i > 10:
        break
    print("Day:", timestamp, "Value:", value, "\n")
    i += 1

Day: 1969-12-31 Value: -8.928451135335273 

Yielding data point #1 from read_fake_data generator.
Day: 1969-12-31 Value: -2.6842722720032413 

Yielding data point #2 from read_fake_data generator.
Day: 1969-12-31 Value: 1.988017533134936 

Yielding data point #3 from read_fake_data generator.
Day: 1969-12-31 Value: -4.318303288053731 

Yielding data point #4 from read_fake_data generator.
Day: 1969-12-31 Value: 1.2756273714792294 

Yielding data point #5 from read_fake_data generator.
Day: 1969-12-31 Value: -3.6213015089109053 

Yielding data point #6 from read_fake_data generator.
Day: 1969-12-31 Value: -5.295121946930241 

Yielding data point #7 from read_fake_data generator.
Day: 1969-12-31 Value: 4.571744935215041 

Yielding data point #8 from read_fake_data generator.
Day: 1969-12-31 Value: -0.9216798216394321 

Yielding data point #9 from read_fake_data generator.
Day: 1969-12-31 Value: -5.16559247678084 

Yielding data point #10 from read_fake_data generator.
Day: 1969-12-31 Val

A minor implementation detail: see that the first `timestamp` and `value` printed, `Day: 1969-12-31 Value: -8.928451135335273`, do not also print `Yielding data point #0 from read_fake_data generator.`; this is because this value was already yielded from the generator and stored temporarily by `groupby` in the background. Because this is a generator, we cannot request it again, so there is no call the the fake data generator here. 

So, `check_anomaly` will be using a `for` loop to essentially call `__next__()` on the iterator contained in `day_data_tuple` (`day_data_tuple[1]`). We can see this in action below:

In [503]:
data = read_fake_data("test", debug=False)
data_day = day_grouper(data)

day_1_tuple = data_day.__next__()

check_anomaly(day_1_tuple)

False

In [506]:
data = read_fake_data("test", debug=False)
data_day = day_grouper(data)

day_tuple = data_day.__next__()

check_anomaly(day_tuple)

datetime.date(1969, 12, 31)

Above we have examples of where we found anomalous dates, and where we did not. In both cases, we are passing in the return tuple from `groupby`, which again returns a tuple of the shape: `(<group key>, <iterable which we can call __next__() on to get group contents>)`.

# 4. `filter` and `map` application
Finally, let's look at how `map` and `filter` can be utilized to make this even more succinct. To start, let's see a simple use case of the `map` function: 

In [4]:
x = [1, 2, 3, 4]
map_obj = map(lambda x: x**2, x)

type(map_obj)

map

In [5]:
type(iter(map_obj))

map

In [7]:
map_obj.__next__()

1

In [8]:
map_obj.__next__()

4

In [9]:
map_obj.__next__()

9

In [10]:
map_obj.__next__()

16

In [11]:
map_obj.__next__()

StopIteration: 

In [12]:
x = [1, 2, 3, 4]
map_obj = map(lambda x: x**2, x)
list(map_obj)

[1, 4, 9, 16]

Above we can clearly see that `map` applies a function against an iterable (in this case simply a list, `x`). Now, of course `map` can also take in a generator as an iterable as well. Let's see if it can take in a `data_day`.

In [26]:
data = read_fake_data("test", debug=False)
data_day = day_grouper(data)

In [27]:
data_day_map = map(check_anomaly, data_day)

In [28]:
data_day_map

In [29]:
data_day_map.__next__()

False

In [30]:
data_day_map.__next__()

datetime.date(1970, 1, 1)

It most definitely can! Keep in mind that `map` is essentially allowing us to simplify passing values from `data_day` to `check_anomaly`; it can be viewed as doing the following:

In [67]:
data = read_fake_data("test", debug=False)
data_day = day_grouper(data)

In [68]:
for i in range(0, 5):
    data_day_tuple = data_day.__next__()
    date_day = check_anomaly(data_day_tuple)
    print(date_day)

False
False
1970-01-02
False
1970-01-04


We can then even combine `map` with `filter`, using `filter` to immediately discard any value of `False`:

In [70]:
data = read_fake_data("test", debug=False)
data_day = day_grouper(data)

data_day_map_filter = filter(None, map(check_anomaly, data_day))

In [71]:
data_day_map_filter.__next__()

datetime.date(1970, 1, 1)

In [72]:
data_day_map_filter.__next__()

datetime.date(1970, 1, 2)

In [73]:
data_day_map_filter.__next__()

datetime.date(1970, 1, 3)

We see that we now only have dates with anomalous values coming through (no more `False`). Again, we can see this working via a simple for loop implementation:

In [74]:
data = read_fake_data("test", debug=False)
data_day = day_grouper(data)

for i in range(0, 5):
    data_day_tuple = data_day.__next__()
    date_day = check_anomaly(data_day_tuple)
    
    if date_day:
        print(date_day)

1970-01-01
1970-01-03
1970-01-04


# 5. `islice` 
The final function to look at is `islice`. It provides a nice way to request a specific number of items from the iterable. It looks like:

In [75]:
data = read_fake_data("test", debug=False)
data_day = day_grouper(data)

anomalous_dates = filter(None, map(check_anomaly, data_day))
next_10_anomalous_dates = islice(anomalous_dates, 10)

In [77]:
next_10_anomalous_dates.__next__()

datetime.date(1969, 12, 31)

In [78]:
list(next_10_anomalous_dates)

[datetime.date(1970, 1, 1),
 datetime.date(1970, 1, 2),
 datetime.date(1970, 1, 3),
 datetime.date(1970, 1, 4),
 datetime.date(1970, 1, 5),
 datetime.date(1970, 1, 6),
 datetime.date(1970, 1, 8),
 datetime.date(1970, 1, 11),
 datetime.date(1970, 1, 13)]

Again, we can see that we can request these values when we need them via `__next__()`, or simply cast to a `list` and get all values at once. Note, this is once again very similar to our looping implementation, only now a `while` loop must be used since we do not know how long it will take to find `10` anomalous values:

In [79]:
data = read_fake_data("test", debug=False)
data_day = day_grouper(data)

num_anomalous_dates = 0
while True:
    data_day_tuple = data_day.__next__()
    date_day = check_anomaly(data_day_tuple)
    
    if date_day:
        print(date_day)
        num_anomalous_dates += 1
    if num_anomalous_dates > 9:
        break

1970-01-01
1970-01-02
1970-01-03
1970-01-06
1970-01-07
1970-01-08
1970-01-09
1970-01-10
1970-01-11
1970-01-12


# 6. `rolling_window_grouper`

In [105]:
def rolling_window_grouper(data, window_size=3600):
    """
    Groups based on a rolling window of data points (instead of grouping for individual days.

    """
    window = tuple(islice(data, 0, window_size))
    while True:
        current_datetime = window[0][0]
        yield (current_datetime, window)
        window = window[1:] + (data.__next__(), )

In [96]:
data = read_fake_data("test", debug=False)

In [97]:
window_size = 10
window = tuple(islice(data, 0, window_size))

In [98]:
window

((datetime.date(1969, 12, 31), 2.812105902775732),
 (datetime.date(1969, 12, 31), 2.0347514988239643),
 (datetime.date(1969, 12, 31), 9.724249881131675),
 (datetime.date(1969, 12, 31), -0.10410381748894187),
 (datetime.date(1969, 12, 31), 1.7615296057616943),
 (datetime.date(1969, 12, 31), -5.084934006233888),
 (datetime.date(1969, 12, 31), -5.680872840329717),
 (datetime.date(1969, 12, 31), -3.0685810315085127),
 (datetime.date(1969, 12, 31), -4.221048380328117),
 (datetime.date(1969, 12, 31), 0.5614347894949309))

In [99]:
window[0][0]

datetime.date(1969, 12, 31)

In [100]:
window[1:]

((datetime.date(1969, 12, 31), 2.0347514988239643),
 (datetime.date(1969, 12, 31), 9.724249881131675),
 (datetime.date(1969, 12, 31), -0.10410381748894187),
 (datetime.date(1969, 12, 31), 1.7615296057616943),
 (datetime.date(1969, 12, 31), -5.084934006233888),
 (datetime.date(1969, 12, 31), -5.680872840329717),
 (datetime.date(1969, 12, 31), -3.0685810315085127),
 (datetime.date(1969, 12, 31), -4.221048380328117),
 (datetime.date(1969, 12, 31), 0.5614347894949309))

In [101]:
(data.__next__(), )

((datetime.date(1969, 12, 31), -1.0769828697393409),)

In [102]:
data.__next__()

(datetime.date(1969, 12, 31), 12.727538639936492)

In [104]:
window[1:] + (data.__next__(),)

((datetime.date(1969, 12, 31), 2.0347514988239643),
 (datetime.date(1969, 12, 31), 9.724249881131675),
 (datetime.date(1969, 12, 31), -0.10410381748894187),
 (datetime.date(1969, 12, 31), 1.7615296057616943),
 (datetime.date(1969, 12, 31), -5.084934006233888),
 (datetime.date(1969, 12, 31), -5.680872840329717),
 (datetime.date(1969, 12, 31), -3.0685810315085127),
 (datetime.date(1969, 12, 31), -4.221048380328117),
 (datetime.date(1969, 12, 31), 0.5614347894949309),
 (datetime.date(1969, 12, 31), 0.27474787498508646))

In [106]:
data = read_fake_data("test", debug=False)
data_window = rolling_window_grouper(data)

In [107]:
date_window_1 = data_window.__next__()

In [110]:
len(date_window_1[1])

3600

In [111]:
date_window_1[1][0]

(datetime.date(1969, 12, 31), -1.9317063809198936)

In [113]:
check_anomaly(date_window_1)

False

In [114]:
date_window_2 = data_window.__next__()
check_anomaly(date_window_2)

False

In [168]:
def test_rolling_window_grouper():
    data = read_fake_data("test_filename")

    data_window = rolling_window_grouper(data)

    anomalous_dates = filter(None, map(check_anomaly, data_window))

    next_100_anomalous_dates = islice(anomalous_dates, 100000)
#     print(list(next_10_anomalous_dates))

In [169]:
%%timeit
test_rolling_window_grouper()

841 ns ± 15.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [172]:
from collections import deque


def rolling_window_grouper_optimized(data, window_size=3600):
    """
    Groups based on a rolling window of data points (instead of grouping for individual days.

    """
    window = deque(islice(data, 0, window_size))
    while True:
        current_datetime = window[0][0]
        yield (current_datetime, window)
        window.popleft()
        window.append((data.__next__(), ))
#         window = window[1:] + (data.__next__(), )

In [173]:
def test_rolling_window_grouper_optimized():
    data = read_fake_data("test_filename")

    data_window = rolling_window_grouper_optimized(data)

    anomalous_dates = filter(None, map(check_anomaly, data_window))

    next_100_anomalous_dates = islice(anomalous_dates, 100000)
#     print(list(next_10_anomalous_dates))

In [174]:
%%timeit
test_rolling_window_grouper_optimized()

841 ns ± 20.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [155]:
data = read_fake_data("test", debug=False)
data_window = rolling_window_grouper_optimized(data)

In [156]:
date_window_1 = data_window.__next__()

In [157]:
len(date_window_1[1])

3600

In [158]:
date_window_1[0]

datetime.date(1969, 12, 31)

In [159]:
date_window_1[1][0]

(datetime.date(1969, 12, 31), 7.228776581819212)

In [160]:
date_window_2 = data_window.__next__()

In [161]:
len(date_window_2[1])

3600

In [162]:
date_window_2[0]

datetime.date(1969, 12, 31)

In [163]:
date_window_2[1][0]

(datetime.date(1969, 12, 31), 2.2356819291893384)

In [164]:
date_window_1

(datetime.date(1969, 12, 31),
 deque([(datetime.date(1969, 12, 31), 2.2356819291893384),
        (datetime.date(1969, 12, 31), 0.677839022192076),
        (datetime.date(1969, 12, 31), -1.7922246739029832),
        (datetime.date(1969, 12, 31), -0.8921900420521836),
        (datetime.date(1969, 12, 31), 1.1548616080302836),
        (datetime.date(1969, 12, 31), -7.327962078838152),
        (datetime.date(1969, 12, 31), 2.7112459135171356),
        (datetime.date(1969, 12, 31), 4.701040221633201),
        (datetime.date(1969, 12, 31), -0.8325776349298027),
        (datetime.date(1969, 12, 31), -2.786167129139573),
        (datetime.date(1969, 12, 31), 0.15826479157685988),
        (datetime.date(1969, 12, 31), 1.0509046505583297),
        (datetime.date(1969, 12, 31), 3.920007088319907),
        (datetime.date(1969, 12, 31), -7.346792973148642),
        (datetime.date(1969, 12, 31), 2.4956732102772725),
        (datetime.date(1969, 12, 31), -6.0281718804747575),
        (datetime.date(1

In [165]:
date_window_2

(datetime.date(1969, 12, 31),
 deque([(datetime.date(1969, 12, 31), 2.2356819291893384),
        (datetime.date(1969, 12, 31), 0.677839022192076),
        (datetime.date(1969, 12, 31), -1.7922246739029832),
        (datetime.date(1969, 12, 31), -0.8921900420521836),
        (datetime.date(1969, 12, 31), 1.1548616080302836),
        (datetime.date(1969, 12, 31), -7.327962078838152),
        (datetime.date(1969, 12, 31), 2.7112459135171356),
        (datetime.date(1969, 12, 31), 4.701040221633201),
        (datetime.date(1969, 12, 31), -0.8325776349298027),
        (datetime.date(1969, 12, 31), -2.786167129139573),
        (datetime.date(1969, 12, 31), 0.15826479157685988),
        (datetime.date(1969, 12, 31), 1.0509046505583297),
        (datetime.date(1969, 12, 31), 3.920007088319907),
        (datetime.date(1969, 12, 31), -7.346792973148642),
        (datetime.date(1969, 12, 31), 2.4956732102772725),
        (datetime.date(1969, 12, 31), -6.0281718804747575),
        (datetime.date(1

In [166]:
date_window_3 = data_window.__next__()
date_window_3

(datetime.date(1969, 12, 31),
 deque([(datetime.date(1969, 12, 31), 0.677839022192076),
        (datetime.date(1969, 12, 31), -1.7922246739029832),
        (datetime.date(1969, 12, 31), -0.8921900420521836),
        (datetime.date(1969, 12, 31), 1.1548616080302836),
        (datetime.date(1969, 12, 31), -7.327962078838152),
        (datetime.date(1969, 12, 31), 2.7112459135171356),
        (datetime.date(1969, 12, 31), 4.701040221633201),
        (datetime.date(1969, 12, 31), -0.8325776349298027),
        (datetime.date(1969, 12, 31), -2.786167129139573),
        (datetime.date(1969, 12, 31), 0.15826479157685988),
        (datetime.date(1969, 12, 31), 1.0509046505583297),
        (datetime.date(1969, 12, 31), 3.920007088319907),
        (datetime.date(1969, 12, 31), -7.346792973148642),
        (datetime.date(1969, 12, 31), 2.4956732102772725),
        (datetime.date(1969, 12, 31), -6.0281718804747575),
        (datetime.date(1969, 12, 31), 0.902278581283916),
        (datetime.date(19

In [167]:
date_window_4 = data_window.__next__()
date_window_4

(datetime.date(1969, 12, 31),
 deque([(datetime.date(1969, 12, 31), -1.7922246739029832),
        (datetime.date(1969, 12, 31), -0.8921900420521836),
        (datetime.date(1969, 12, 31), 1.1548616080302836),
        (datetime.date(1969, 12, 31), -7.327962078838152),
        (datetime.date(1969, 12, 31), 2.7112459135171356),
        (datetime.date(1969, 12, 31), 4.701040221633201),
        (datetime.date(1969, 12, 31), -0.8325776349298027),
        (datetime.date(1969, 12, 31), -2.786167129139573),
        (datetime.date(1969, 12, 31), 0.15826479157685988),
        (datetime.date(1969, 12, 31), 1.0509046505583297),
        (datetime.date(1969, 12, 31), 3.920007088319907),
        (datetime.date(1969, 12, 31), -7.346792973148642),
        (datetime.date(1969, 12, 31), 2.4956732102772725),
        (datetime.date(1969, 12, 31), -6.0281718804747575),
        (datetime.date(1969, 12, 31), 0.902278581283916),
        (datetime.date(1969, 12, 31), -0.14232924590620702),
        (datetime.date